### ISSUE TITLE
Usage in Scancode Workbench for arbitrary expressions?

### ISSUE THREAD

**Issue by pombredanne** [lkoskela/license-expressions#98]
Hi Lasse!

Thanks for this library which works nicely with expression that do not use SPDX license keys!

@OmkarPh wants to use it in Scancode Workbench. There we use SPDX expressions and also other license keys that are not from SPDX.

I maintain the Python library at https://github.com/nexB/license-expression/ and it would be awesome if we can have similar features in your library (and we will try to help as much as we can)

Some other features of interest:
- having a function to extract the flat list of license keys and exceptions
- being to render an expression with a string template, for instance to have an href/link to each license in the expression
- optionally validating against a provided list of license and expression keys

**Comment by OmkarPh** [comment]
yeah, the option to get a flat list of keys & exceptions would be a great help

**Comment by OmkarPh** [comment]
Hi @lkoskela 
Is there any way to parse like this 
`apache-1.1` -> `{ license: 'apache-1.1' }` instead of the default `{ license: 'Apache-1.1' }`
`lgpl-2.1` -> `{ license: 'lgpl-2.1' }` instead of  default `{ license: 'LGPL-2.1-only' }`

**Comment by lkoskela** [comment]
@OmkarPh 
> Hi @lkoskela
> Is there any way to parse like this
> `apache-1.1` -> `{ license: 'apache-1.1' }` instead of the default `{ license: 'Apache-1.1' }`
> `lgpl-2.1` -> `{ license: 'lgpl-2.1' }` instead of default `{ license: 'LGPL-2.1-only' }`

https://github.com/lkoskela/license-expressions/pull/105 makes the "upgrade" configurable (and changes the default behaviour to _not_ upgrading). Does that match what you're looking for?

**Comment by lkoskela** [comment]
@OmkarPh 
> yeah, the option to get a flat list of keys & exceptions would be a great help

This would be quite doable since `license-expressions` already contains that data (pulled from [SPDX Workgroup's data files](https://github.com/spdx/license-list-data/tree/main/json) at build time).

Do you have an idea of what the API would ideally look like from your point of view?

**Comment by OmkarPh** [comment]
> @OmkarPh
> 
> > Hi @lkoskela
> > Is there any way to parse like this
> > `apache-1.1` -> `{ license: 'apache-1.1' }` instead of the default `{ license: 'Apache-1.1' }`
> > `lgpl-2.1` -> `{ license: 'lgpl-2.1' }` instead of default `{ license: 'LGPL-2.1-only' }`
> 
> #105 makes the "upgrade" configurable (and changes the default behaviour to _not_ upgrading). Does that match what you're looking for?

yeah this looks like what we expected
At AboutCode, we have licenses other than SPDX too, so we want to parse the components as they are without any changes
or SPDX verification
Can you provide a parse option that suits this

**Comment by OmkarPh** [comment]
> @OmkarPh
> 
> > yeah, the option to get a flat list of keys & exceptions would be a great help
> 
> This would be quite doable since `license-expressions` already contains that data (pulled from [SPDX Workgroup's data files](https://github.com/spdx/license-list-data/tree/main/json) at build time).
> 
> Do you have an idea of what the API would ideally look like from your point of view?

we just need a list, nothing else, for example `parse('gpl-2.0-plus WITH ada-linking-exception', { some option here })`
should return an array of strings i.e. `["gpl-2.0-plus", "ada-linking-exception"]`

**Comment by lkoskela** [comment]
@OmkarPh I published 0.6.0, which includes the `--upgrade` option to enable "upgrading" from a deprecated license (such as `GPL-3.0`) to the respective non-deprecated version (e.g. `GPL-3.0-only`).

**Comment by lkoskela** [comment]
> hi @lkoskela , sorry for late reply i tried latest 0.6.0, but still am getting results like earlier
> 
> ```
> console.log(parse('GPL-3.0+'));
> console.log(parse('apache-1.1', { strictSyntax: false, upgradeGPLVariants: false }));
> console.log(parse('lgpl-2.1'));
> ```
> 
> ![image](https://user-images.githubusercontent.com/48476025/229194628-08006a07-f2cc-4c51-902f-a05d8184f8c6.png)
> 
> Expected something like this:
> 
> ```
> { license: 'GPL-3.0+' }
> { license: 'apache-1.1' }
> { license: 'lgpl-2.1' }
> ```

Right. It looks like our thoughts went in different directions regarding the desired behavior. To try and clarify things, which of the following behaviours would you like to have?

1. Disabling _ambiguous_ corrections/transformations. (For example, `LGPL-2.1` would not be coerced into neither `LGPL-2.1-only` nor `LGPL-2.1-or-later` because we don't know for certain which of the two did the licensor intend when they wrote down "LGPL-2.1".)
2. Disabling _unambiguous_ corrections/transformations. (For example, For example, `apache-1.1` would not be uppercased into its official form of `Apache-1.1`, `GPL-3.0+` would not be coerced into `GPL-3.0-or-later`, and `GPLv3` would not be coerced into `GPL-3.0`.)
3. Disabling _any_ corrections/transformations whatsoever.
4. Something else?

The `upgradeGPLVariants` option added in version 0.6.0 actually only does 1. (Although its name implies it might be doing a bit of 2. as well, correcting the common misspellings/typos of GPL family license identifiers)

I'm now thinking you want 3. i.e. no transformations on the license identifiers whatsoever. Do correct me if I'm wrong.

**Comment by OmkarPh** [comment]
> I'm now thinking you want 3. i.e. no transformations on the license identifiers whatsoever. Do correct me if I'm wrong.

Yes that's correct 
Also, you can make this behaviour optional, keep the default behaviour as it suits for you
